### DISCUSSION TITLE
Using an async function in describe instead of beforeAll()?

### DISCUSSION THREAD

**Discussion by katangafor** [vitest-dev/vitest#4858] [Q&A] [answered]
When using beforeAll(), the tests don't have access to any variables created inside the beforeAll callback.

Is there anything wrong with doing something like this?

```
describe("awesome test suite", async () => {
  const fakeClient = await prisma.client.create({
    data: {
      name: "hello mr. client"
    }
  })

  test("client has an id", async () => {
    expect(fakeClient.id).toBeDefined()
  })
})
```

**Comment by sheremet-va** [comment]
Vitest runs tests in two phases:

1. Collect phase -- all tests are collected, but no `before`/`after`/`test` functions are called yet
2. Execute phase -- all collected tests and hooks are executed in order

If you don't use a hook, then `prisma` is called during the collection phase. This doesn't really matter unless you want your collection phase to be fast. For example, when using https://github.com/vitest-dev/vitest/pull/4240 in the future (IDE integrations run your code and create `prisma` instances, but don't call `after` hooks so they are never cleared which might case issues).

**Comment by SimonEast** [comment]
I've just been attempting the same thing, however I've experienced weird issues regarding the ordering of execution, presumably because of the different phases that @sheremet-va described. For example when I have:

Describe 1  
\- Test 1a
\- Test 1b
Describe 2
\- Test 2a
\- Test 2b

While you may expect everything to run in sequence, instead what happens is Describe 1 and Describe 2 are run first, before all of the tests, and then the tests are run 1a, 1b, 2a, 2b.

In my case a shared/global variable was getting modified by Describe 2 and affecting tests 1a and 1b, even though it comes later in the script.

So I think I will have to refactor to _define the variable inside the describe_ (so they can be accessed by all tests) but use the `beforeAll()` method to populate it. Then things should run in the expected order, like this:

\- Describe 1  
\- Describe 2
\- beforeAll() 1
\- Test 1a
\- Test 1b
\- beforeAll() 2
\- Test 2a
\- Test 2b

**Comment by alexey-sh** [comment]
```
describe('DB', () => {
    let posts: Post[] = [];
    beforeAll(async () => {
        const loadedPosts = await prisma.post.findMany({ where: {}});
        posts.push(...loadedPosts);
    });
    describe('lol', () => {
        test('kek', () => {
            console.log(posts.length);
            expect(true).toBe(true)
        })
        test.each(posts)('Test $external_id', async (post) => {
            const content = getPostText(post.external_id, post.data, post.data_source);
            await expect(content).toMatchFileSnapshot(`./db-snapshots/${post.data_source}_${post.external_id}.txt`);
        });
    });
});
```

`test.each` does not handled. So the `beforeAll` approach does not work for me. However i see in console this:
```
stdout | test/content-extractor.test.ts > DB > lol > kek
224
```

**Reply by SimonEast** [reply]
Does it work if you put the `beforeAll()` inside the `describe('lol'...` section?

And is `getPostText()` an async function? If so, maybe you need to `await` it *before* you call `expect()`.

**Reply by hi-ogawa** [reply, marked as answer]
This is expected. As explained in https://github.com/vitest-dev/vitest/discussions/4858#discussioncomment-8005117,
- collection phase: run `describe` and `test.each(posts)`
- execution phase: run `() => { ... }` functions from `beforeAll(() => { ... })` and `test('kek', () => { ... })`, `test.each(posts)('...', () => { ... })`

**Reply by alexey-sh** [reply]
> Does it work if you put the `beforeAll()` inside the `describe('lol'...` section?
> 
> And is `getPostText()` an async function? If so, maybe you need to `await` it _before_ you call `expect()`.

no it does not.
getPostText is not async
